Clean up some DataBox code

kidder opened the pull request:
## Proposed changes

My forthcoming PR refactoring DataBox was getting a bit long, so I have factored out a few simple commits.
The next PR will replace the implementation of `DataBoxLeaf` with one that does not use `Deferred`

The biggest change in this PR is to the implementations of `db::apply` and `db::mutate_apply` so that their implementations are similar.

The other commits are of several types:
- Clean up documentation
- Remove or simplify typelists 

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Update documentation of structs marking DataBox tags

kidder committed:
Add missing cond-endcond around forward decls

kidder committed:
Simplify and update documentation for DataBox TagTraits

kidder committed:
Update documentation for TagName

kidder committed:
Update documentation for PrefixHelpers

kidder committed:
Remove unnecessary uses of db::split_tag

kidder committed:
Change Boost_INCLUDE_DIR to Boost_INCLUDE_DIRS

Necessary to compile on my laptop, and also what we report in SetupBoost.cmake

kidder committed:
Remove unused split_tag and prefix_tag_wraps_specified_tag

kidder committed:
Remove uses of db::item_type

kidder committed:
Put const_item_type and item_type into namespace db::detail

They are now used only internally by DataBox and Domain/InterfaceHelpers.
Also rename namespace db::DataBox_detail to db::detail.

kidder committed:
Fix some DataBox includes

kidder committed:
Simplify db::apply and db::mutate_apply

Use if constexpr and make their implmentations more consistent.

kidder committed:
Simplify construction of typelists with subitems in DataBox

kidder committed:
Change how lists of compute items to be reset are found.

Remove duplicates from list and don't reset some items twice.  This
speeds up db::mutate for a large number of compute tags.

nilsdeppe commented on src/DataStructures/DataBox/Tag.hpp:
```diff
@@ -6,96 +6,141 @@
+ * from a DataBox.
+ *
+ * A compute tag contains a member named `function` that is either a
+ * function pointer, or a static constexpr function. The compute tag
```
I think it only needs to be `static` if it is a function, but `static constexpr` if it is a variable pointing to a function (function pointer)

nilsdeppe commented on src/DataStructures/DataBox/TagTraits.hpp:
```diff
@@ -18,45 +18,52 @@ struct ComputeTag;
 /// \endcond
 
+/// \ingroup DataBoxGroup
+/// \brief True if `Tag` derives from db:: ComputeTag.
```
is the space between `db::` and `ComputeTag` intentional?

nilsdeppe commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1309,49 +1309,54 @@ namespace detail {
+constexpr void error_function_not_callable() noexcept {
+  static_assert(std::is_same_v<Func, void>,
+                "The function is not callable with the expected arguments.  "
+                "See the first template parameter for the function type and "
```
`template parameter...` -> `template parameter of error_function_not_callable...`

nilsdeppe commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1309,49 +1309,54 @@ namespace detail {
+    DataBoxTags /*meta*/, tmpl::list<TagsToRetrieve...> /*meta*/) noexcept {
+  static_assert(
+      (tag_is_retrievable_v<TagsToRetrieve, DataBox<DataBoxTags>> and ...),
+      "A desired tag is not in the DataBox.  See the first template "
```
`template argument...` -> `template argument of check_tags_are_in_databox...`

nilsdeppe commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -120,155 +120,20 @@ class DataBoxLeaf {
-  using f = tmpl::list<>;
+template <typename... Tags>
+struct expand_subitems_impl<tmpl::list<Tags...>> {
+  using type = tmpl::append<append_subitem_tags<Tags>...>;
```
:drooling_face:

nilsdeppe reviewed (changes requested):
noice! Go ahead and squash immediately :)

kidder commented:
@wthrowe or @nilsleiffischer could you provide a second review for this PR?

wthrowe commented on src/DataStructures/DataBox/Tag.hpp:
```diff
@@ -6,96 +6,141 @@
+ * tag that derives from it appended by `Base`.  Alternatively, if the simple
+ * tag is templated with a type used to determine its `type` type alias, the
+ * base tag can have the same name as the simple tag template with an empty
+ * template parameter list.
```
It does need a defaulted parameter, but that may be too much detail to get into here since anyone trying to implement this will look up an example anyway.

wthrowe commented on src/DataStructures/DataBox/Tag.hpp:
```diff
@@ -6,96 +6,141 @@
 /*!
  * \ingroup DataBoxGroup
- * \brief Marks an item as being a prefix to another tag
+ * \brief Marks a struct as a prefix tag by inheriting from this.
```
`Marks` -> `Mark`

wthrowe commented on src/DataStructures/DataBox/Tag.hpp:
```diff
@@ -6,96 +6,141 @@
 /*!
  * \ingroup DataBoxGroup
- * \brief Marks a DataBoxTag as being a compute item that executes a function
+ * \brief Marks a struct as a compute tag by inheriting from this.
```
`Marks` -> `Mark`

nilsvu commented:
Sure, Monday at the latest

wthrowe commented on src/DataStructures/DataBox/TagTraits.hpp:
```diff
@@ -18,45 +18,52 @@ struct ComputeTag;
+ * \see is_compute_tag_v ComputeTag
  */
 template <typename Tag, typename = std::nullptr_t>
 struct is_compute_tag : std::false_type {};
-/// \cond HIDDEN_SYMBOLS
+/// \cond
 template <typename Tag>
 struct is_compute_tag<Tag, Requires<std::is_base_of_v<db::ComputeTag, Tag>>>
     : std::true_type {};
 /// \endcond
```
Not needed for this PR, but this would be simpler as
```c++
template <typename Tag>
struct is_compute_tag : std::bool_constant<std::is_base_of_v<db::ComputeTag, Tag>> {};
```

wthrowe commented on src/DataStructures/DataBox/Tag.hpp:
```diff
@@ -6,96 +6,141 @@
+ *
+ * \warning Do not derive a base tag from anything besides db::BaseTag.
+ *
+ * \see DataBoxGroup BaseTag SimpleTag
```
BaseTag is see-alsoing itself.

wthrowe commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -799,13 +799,15 @@ db::DataBox<tmpl::list<Tags...>>::reset_compute_items_after_mutate(
-                          std::is_same<tmpl::pin<ComputeItemsToReset>,
-                                       tmpl::get_source<tmpl::_1>>>...>,
-                      tmpl::get_destination<tmpl::_1>>;
+  using current_tags_to_reset = tmpl::list<ComputeItemsToReset...>;
```
Should this also contain the tags reset in previous iterations of this reset procedure?

kidder replied:
Yes it should in order to be more efficient, but it is not necessary.  This function will be changed in my next PR, so if it is okay, I'd prefer to make this improvement then

wthrowe replied:
OK.

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1621,102 +1364,42 @@ constexpr bool has_return_tags_and_argument_tags_v =
-                                      ArgumentTags{},
-                                      std::forward<Args>(args)...);
+    Args&&... args) noexcept {
+  mutate_apply<typename std::decay_t<F>::return_tags,
```
If I understand correctly you're removing the possibility that `mutate_apply` returns values. We're probably not using that functionality anywhere right now, but it could be useful at some point, don't you think?

kidder replied:
no, detail::mutate_apply doesn't return anything (it just calls db::mutate)

nilsvu reviewed (commented):
Not a review, just noticed this

kidder commented:
@nilsdeppe @wthrowe pushed fixups

nilsdeppe commented on src/DataStructures/DataBox/TagTraits.hpp:
```diff
@@ -21,18 +21,15 @@ namespace db {
-    : std::true_type {};
-/// \endcond
+struct is_compute_tag
+    : std::bool_constant<std::is_base_of_v<db::ComputeTag, Tag>> {};
```
isn't this just `: std::is_base_of<db::ComputeTag, Tag>`? No need to convert to a bool constant I think

nilsdeppe reviewed (commented):
I'm happy with things. One tiny suggestion, feel free to squash once @wthrowe gives you the go-ahead :)

wthrowe commented:
Looks good once you make @nilsdeppe's correction to my correction.

kidder committed:
Include StlStreamDeclarations.hpp in MakeString.hpp

Removing the include of DataBoxTag.hpp in Domain/Tags.hpp removes a transitive
include of StlStreamDeclarations.hpp causing executables to fail to compile with
the error: "call to function 'operator<<' that is neither visible in the template
definition nor found by ADL" in the friend operator<< that returns by value.

This is a temporary fix until we fix includes and factor StdHelpers.hpp to avoid
including most of the STL when it StdHelpers.hpp is included.

kidder committed:
Fix clang-tidy error in FirstOrderEllipticSolutionsHelpers.hpp

Do not std::move fluxes as it is taken by reference

kidder commented:
@wthrowe @kidder squashed and rebased.  There are two new commits to deal with CI failures after rebasing.  The first is "Include StlStreamDeclarations.hpp in MakeString.hpp " and the second is the "Fix clang-tidy ..."

wthrowe commented on tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:
```diff
@@ -80,6 +78,8 @@ void verify_solution(
       },
       sources_args);
+  auto div_fluxes = divergence(std::move(fluxes), mesh,
+                               coord_map.inv_jacobian(logical_coords));
```
`divergence` takes its argument by reference, so the `std::move` shouldn't be there at all.

kidder commented:
@wthrowe squashed and rebased

kidder commented:
@wthrowe rebased and squashed the clang-tidy commit

wthrowe commented on tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:
```diff
@@ -70,8 +70,8 @@ void verify_solution(
-  auto div_fluxes = divergence(std::move(fluxes), mesh,
-                               coord_map.inv_jacobian(logical_coords));
+  auto div_fluxes =
+      divergence(fluxes, mesh, coord_map.inv_jacobian(logical_coords));
```
Commit message is outdated.

kidder commented:
@wthrowe thanks, fixed the commit message
